src/transaction/lock.rs

```diff
@@ -21,67 +21,79 @@ pub async fn resolve_locks(
-    let expired_locks = locks.into_iter().filter(|lock| {
-        let expired = ts.physical - Timestamp::from_version(lock.lock_version).physical
-            >= lock.lock_ttl as i64;
-        if !expired {
-            has_live_locks = true;
-        }
-        expired
-    });
+    let locks_len = locks.len();
 
-    // records the commit version of each primary lock (representing the status of the transaction)
-    let mut commit_versions: HashMap<u64, u64> = HashMap::new();
-    let mut clean_regions: HashMap<u64, HashSet<RegionVerId>> = HashMap::new();
-    for lock in expired_locks {
-        let primary_key: Key = lock.primary_lock.into();
-        let region_ver_id = pd_client.region_for_key(&primary_key).await?.ver_id();
-        // skip if the region is cleaned
-        if clean_regions
-            .get(&lock.lock_version)
-            .map(|regions| regions.contains(&region_ver_id))
-            .unwrap_or(false)
-        {
-            continue;
+    struct RichLock {
+        primary_key: Key,
+        region: Region,
+        lock_version: u64,
+        txn_size: u64,
+    }
+    let mut expired_locks = vec![];
+    for lock in locks {
+        let expired = ts.physical - Timestamp::from_version(lock.clone().lock_version).physical
+            >= lock.clone().lock_ttl as i64;
+        if expired {
+            let primary_key: Key = lock.primary_lock.into();
+            let region = pd_client.region_for_key(&primary_key).await?;
+            expired_locks.push(RichLock {
+                primary_key,
+                region,
+                lock_version: lock.lock_version,
+                txn_size: lock.txn_size,
+            })
         }
+    }
 
-        let commit_version = match commit_versions.get(&lock.lock_version) {
-            Some(&commit_version) => commit_version,
-            None => {
-                let commit_version = requests::new_cleanup_request(primary_key, lock.lock_version)
-                    .execute(pd_client.clone())
-                    .await?;
-                commit_versions.insert(lock.lock_version, commit_version);
-                commit_version
-            }
-        };
+    let has_live_locks = locks_len > expired_locks.len();
 
-        let cleaned_region = resolve_lock_with_retry(
-            lock.key.into(),
-            lock.lock_version,
+    let grouped: Vec<((RegionVerId, u64), Vec<RichLock>)> = expired_locks
+        .into_iter()
+        .group_by(|rich_lock| {
+            (
+                rich_lock.region.ver_id().clone(),
+                rich_lock.lock_version.clone(),
+            )
+        })
+        .into_iter()
+        .map(|(k, v)| (k, v.collect()))
+        .collect();
+
+    for ((_region_ver_id, lock_version), locks) in grouped {
+        let is_large_txn = locks.iter().any(|lock| lock.txn_size >= 16);
+
+        let commit_version =
+            requests::new_cleanup_request(locks.get(0).unwrap().primary_key.clone(), lock_version)
```

Comment:
Why is only the first lock in a region cleaned up?

Reply:
> Why is only the first lock in a region cleaned up?

https://github.com/Renkai/client-rust/blob/3ab4252e1dce5d1a32042ca7fb028329a5a1426e/src/transaction/lock.rs#L49

In the original version, every lock_version will do a region clean up only once. Here the locks are already grouped by lock version, so only the first grouped lock do a region clean up behave exactly the same as original.

Reply:
Ah, I got what you mean. The idea looks good. Yet there is something to refine.
For every lock with the same `(region, lock_version)`, the `txn_size` is the same. So actually we don't need to save the keys of the large transactions.
And there seems to be lots of clones, could you try to reduce clones?